### PR TITLE
T0 model and  dataset

### PR DESCRIPTION
**Pull request by MaximumEntropy**
# What does this PR do ?

Adds the T0 Dataset and Model classes.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Adds the T0 dataset and model class files
- Repurposes the generic fine-tuning run script for T0

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/common/data/dataset.py [hunk 1]
```diff
@@ -176,3 +176,99 @@ def random_generator(datasets, **kwargs):
...
+
+    def __getitem__(self, idx):
+        # Get the dataset we want to sample from
+        dataset_index = self._get_dataset_index(idx)
+
+        # Get the index of the sample we want to fetch from the dataset
+        sample_idx = self.dataset_index[dataset_index]
+
+        # Sample index -> shuffled sample index
+        shuffled_sample_idx = self.permuted_dataset_indices[dataset_index][sample_idx]
```
Comment (yaoyu-33) [inline nemo/collections/common/data/dataset.py:264]: `sample_idx` here could get out of boundary and should be in the if below?

Comment (MaximumEntropy) [reply nemo/collections/common/data/dataset.py:264]: Oh yes! Thats a good point. Will fix this.

Comment (MaximumEntropy) [reply nemo/collections/common/data/dataset.py:264]: Good point! Will fix this.

File: examples/nlp/language_modeling/megatron_t5_seq2seq_finetune.py [hunk 2]
```diff
@@ -106,6 +107,13 @@ def main(cfg) -> None:
             override_config_path=t5_cfg,
             save_restore_connector=NLPSaveRestoreConnector(),
         )
+    elif hasattr(cfg.model.data.train_ds, 'file_names'):
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/megatron_t5_seq2seq_finetune.py:110]: I don't think we have a better way at least right now if we want to share the same fine-tuning script.

File: nemo/collections/common/data/dataset.py (hunk 1 above)
Comment (MaximumEntropy) [reply nemo/collections/common/data/dataset.py]: Fixed.

**Review by ericharper — DISMISSED** [review]
LGTM. Thanks!

**Review by ericharper — APPROVED** [review]
Re-approving.

**Pull request merged**